### ISSUE TITLE
Feature Request - Add Indexing to the tree-based collections

### ISSUE THREAD

**Issue by JohnnyNonsense** [google/guava#2957]
Here is an example where someone modified the TreeSet and TreeMap to include indexing functionality:
https://code.google.com/archive/p/indexed-tree-map/

I find this to be very useful, and is easily added to any binary search tree.  Just include the weight of each subtree, and update the weights as necessary.  Use the weights to find nodes by index, and vice-versa.

**Comment by liach** [comment]
It is more similar to the existing tree multi set.

**Comment by JohnnyNonsense** [comment]
I found a couple of more examples.  In javolution.org, they have a FastSortedTable, which supports indexing.  No idea if it uses a tree, but it is basically a multiset.
http://javolution.org/apidocs/index.html

Then the Goldman Sachs / Eclipse Collections has something called TreeSortedSet, and UnmodifiableSortedSet.  These have indexOf, and forEach using a range of indices.
https://www.eclipse.org/collections/javadoc/9.0.0/index.html?overview-summary.html

**Comment by jrtom** [comment]
What specific use cases do you have in mind for these data structures--that is, what problems are you trying to solve that require them?

Some (pseudo-)code snippets might be helpful to get your ideas across.

**Comment by jrtom** [comment]
@liach I understand that they're sorted _sets_, or maps (not lists), with the added ability to retrieve an item given its position in the ordering.

The question is: what kinds of problems are you trying to solve that require this data structure?

**Comment by liach** [comment]
Like some sort of alternative to `MinMaxPriorityQueue`?

**Comment by Maaartinus** [comment]
- uniformly random selection
- median and other quantiles

**Comment by liach** [comment]
agree with @maaartinus. you can use tgis extended ability to create stemplots of any size, too.

**Comment by lowasser** [comment]
Note Guava already has https://google.github.io/guava/releases/21.0/api/docs/com/google/common/math/Quantiles.html for that sort of operation.

**Comment by liach** [comment]
@lowasser Your alternative may be very inefficient if the data set is modified frequently. What's more, it only supports Doubles.

**Comment by lowasser** [comment]
That's correct, but what I'm still not hearing is business use cases for which this is clearly the best solution, for which the data set _is_ modified frequently or _doesn't_ consist of doubles.  Do you _have_ a data set that is modified frequently?  What is in it?  Is it customer ID numbers?  Costs for pencils?  Please be more specific than "numbers" or "objects" or the like.

Do you have an actual application for which you need to compute the median and other quantiles, and that's a significant performance bottleneck that needs improvement?  What are you computing the median of?  How many data points do you have, to a rough order of magnitude?

**Comment by lowasser** [comment]
(Mostly right now I'm just trying to understand the details asked for here: https://github.com/google/guava/wiki/HowToContribute#ubiquity-provide-concrete-use-cases )

**Comment by liach** [comment]
It is a better alternative to GapList here https://dzone.com/articles/gaplist-lightning-fast-list
